os/net/mac/csma/csma-output.c

```diff
@@ -200,6 +200,8 @@ send_one_packet(void *ptr)
       case RADIO_TX_OK:
         if(is_broadcast) {
           ret = MAC_TX_OK;
+        } else if (!CSMA_REQUEST_ACK) {
```

Comment:
One more thing: I wonder if instead of this added condition, we shouldn't just replace the variable `is_broadcast` with a new `is_ack_requested`. Would make more sense w.r.t handling of packet pending prior to tx (code block above).

Reply:
Makes perfect sense!

One thing though, I don't get the point of the second condition here:

```c
if(NETSTACK_RADIO.receiving_packet() ||
   (!is_broadcast && NETSTACK_RADIO.pending_packet())) {
```

We don't transmit if the radio just received a packet and the outgoing packet is unicast?  What's the logic behind it? Is the idea that if we received a unicast packet (and not done processing) it's considered a collision since we should be responding with an ACK first?



Where you thinking of something like this? 

```diff
diff --git a/os/net/mac/csma/csma-output.c b/os/net/mac/csma/csma-output.c
index 9c88d569d..9c70ea8c3 100644
--- a/os/net/mac/csma/csma-output.c
+++ b/os/net/mac/csma/csma-output.c
@@ -168,27 +168,28 @@ backoff_period(void)
 static int
 send_one_packet(void *ptr)
 {
+  int expect_ack;
   int ret;
   int last_sent_ok = 0;
 
+  expect_ack = !packetbuf_holds_broadcast() || CSMA_REQUEST_ACK;
+
   packetbuf_set_addr(PACKETBUF_ADDR_SENDER, &linkaddr_node_addr);
-  packetbuf_set_attr(PACKETBUF_ATTR_MAC_ACK, 1);
+  packetbuf_set_attr(PACKETBUF_ATTR_MAC_ACK, expect_ack);
 
   if(NETSTACK_FRAMER.create() < 0) {
     /* Failed to allocate space for headers */
     LOG_ERR("failed to create packet\n");
     ret = MAC_TX_ERR_FATAL;
   } else {
-    int is_broadcast;
     uint8_t dsn;
     dsn = ((uint8_t *)packetbuf_hdrptr())[2] & 0xff;
 
     NETSTACK_RADIO.prepare(packetbuf_hdrptr(), packetbuf_totlen());
 
-    is_broadcast = packetbuf_holds_broadcast();
 
     if(NETSTACK_RADIO.receiving_packet() ||
-       (!is_broadcast && NETSTACK_RADIO.pending_packet())) {
+       (expect_ack && NETSTACK_RADIO.pending_packet())) {
 
       /* Currently receiving a packet over air or the radio has
          already received a packet that needs to be read before
@@ -198,7 +199,7 @@ send_one_packet(void *ptr)
 
       switch(NETSTACK_RADIO.transmit(packetbuf_totlen())) {
       case RADIO_TX_OK:
-        if(is_broadcast) {
+        if(!expect_ack) {
           ret = MAC_TX_OK;
         } else {
           rtimer_clock_t wt;
```

Reply:
The idea I believe is that if there is something pending we might have the Rx fifo full and be not able to receive the ACK.

At first glance the patch above looks good yes!